### PR TITLE
Better load interoperability

### PR DESCRIPTION
**Pull request by jfalcou**
Made load a bit better at being overloaded

### CODE REVIEW

File: include/eve/detail/function/simd/common/load.hpp [hunk 1]
```diff
@@ -11,11 +11,95 @@
...
+  template<data_source Ptr, typename Cardinal = void>
+  struct loaded_wide : as_wide<typename pointer_traits<Ptr>::value_type, Cardinal>
+  {};
+
+  template<data_source Ptr>
+  struct loaded_wide<Ptr> : as_wide<typename pointer_traits<Ptr>::value_type>
+  {};
+
+  template<data_source Ptr, typename Cardinal = void>
+  using loaded_wide_t = typename loaded_wide<Ptr,Cardinal>::type;
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/common/load.hpp:35]: I'd prefer to call it `wide_value_type_t` tbh

File: include/eve/memory/pointer.hpp [hunk 2]
```diff
@@ -8,11 +8,71 @@
...
+    using base = typename pointer_traits<Ptr>::value_type;
+    static constexpr bool value = std::same_as<T,base>;
+  };
+
+  template<typename T, typename... Ptrs>
+  struct dereference_as<T, kumi::tuple<Ptrs...>>
+  {
+    using base = kumi::tuple<typename pointer_traits<Ptrs>::value_type...>;
+    static constexpr bool value = std::same_as<T,base>;
+  };
```
Comment (DenisYaroshevskiy) [inline include/eve/memory/pointer.hpp:74]: Hmm, this only allows for kumi::tuple. But w/e

File: include/eve/module/real/core/function/regular/generic/load.hpp [hunk 3]
```diff
@@ -44,149 +46,57 @@ namespace eve::detail
...
 
     if constexpr( !std::is_pointer_v<Ptr> )
     {
-      static constexpr bool is_aligned_enough = Cardinal() * sizeof(e_t) >= Ptr::alignment();
+      static constexpr bool is_aligned_enough = c_t() * sizeof(e_t) >= Ptr::alignment();
 
       if constexpr (!sanitizers_are_on && is_aligned_enough)
       {
-        auto that = eve::unsafe(eve::load)(ptr, Cardinal{});
+        auto that = eve::unsafe(eve::load)(ptr, c_t{});
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/load.hpp:85]: `as` is lost

File: include/eve/module/real/core/function/regular/generic/load.hpp [hunk 4]
```diff
@@ -44,149 +46,57 @@ namespace eve::detail
...
+        return eve::load[cond](ptr.get(), c_t{});
       }
     }
     else
     {
       // If the ignore/keep is complete we can jump over if_else
       if constexpr( C::is_complete )
       {
-        if constexpr(C::is_inverted)  { return eve::load(ptr, Cardinal{});  }
+        if constexpr(C::is_inverted)  { return eve::load(ptr, c_t{});  }
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/load.hpp:99]: In all of these

File: include/eve/module/real/core/function/regular/generic/load.hpp [hunk 5]
```diff
@@ -240,69 +133,33 @@ namespace eve::detail
...
+  template<relative_conditional_expr C, data_source Ptr, typename Pack>
+  DISABLE_SANITIZERS Pack load_ ( EVE_SUPPORTS(cpu_)
+                                , C const& cond, unsafe_type const&
+                                , eve::as_<Pack> const& tgt, Ptr ptr
+                                ) noexcept
   {
-    using T = std::remove_cvref_t<decltype(*ptr)>;
-    return eve::unsafe(eve::load[cond])(ptr, expected_cardinal_t<T>{});
-  }
+    if constexpr(sanitizers_are_on)
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/load.hpp:145]: and not `ignore_none`

File: include/eve/arch/cpu/logical_wide.hpp [hunk 6]
```diff
@@ -113,7 +114,7 @@ namespace eve
     //! Construction is done piecewise unless the @iterator{s} are @raiterator{s}.
     template<std::input_iterator Iterator>
     EVE_FORCEINLINE explicit logical(Iterator b, Iterator e) noexcept
-                  : storage_base(detail::load(eve::as_<logical>{}, b, e))
+                  : storage_base(load(ignore_none, safe, eve::as_<logical>{}, b, e))
```
[the remark this answers is not in the document]
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/logical_wide.hpp:117]: Don't do the change fo contigious iterators

File: include/eve/algo/ptr_iterator.hpp [hunk 7]
```diff
@@ -21,8 +21,9 @@ namespace eve::algo
   template <typename T, typename Cardinal>
   struct unaligned_ptr_iterator : operations_with_distance
   {
-    using cardinal = Cardinal;
+    using cardinal        = Cardinal;
     using wide_value_type = eve::wide<std::remove_const_t<T>, cardinal>;
+    using value_type      = typename wide_value_type::value_type;
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/algo/ptr_iterator.hpp:26]: Fine

File: include/eve/algo/ptr_iterator.hpp [hunk 8]
```diff
@@ -34,20 +35,21 @@ namespace eve::algo
       return aligned_ptr_iterator<T, cardinal>{eve::previous_aligned_address(ptr, cardinal{})};
     }
 
+    decltype(auto) operator*() const { return *ptr; }
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/algo/ptr_iterator.hpp:38]: Sure

File: include/eve/algo/ptr_iterator.hpp [hunk 9]
```diff
@@ -70,10 +72,10 @@ namespace eve::algo
...
   struct aligned_ptr_iterator : operations_with_distance, forward_to_unaligned
   {
-    using cardinal = Cardinal;
-    using wide_value_type = eve::wide<std::remove_const_t<T>, cardinal>;
-
-    using aligned_ptr_type = eve::aligned_ptr<T, Cardinal>;
+    using cardinal          = Cardinal;
+    using aligned_ptr_type  = eve::aligned_ptr<T, Cardinal>;
+    using value_type        = typename aligned_ptr_type::value_type;
+    using wide_value_type   = eve::wide<std::remove_const_t<T>, cardinal>;
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/algo/ptr_iterator.hpp:78]: Done

File: include/eve/arch/cpu/logical_wide.hpp [hunk 10]
```diff
@@ -126,7 +127,7 @@ namespace eve
     //! Constructs a eve::logical from a SIMD compatible pointer
     template<simd_compatible_ptr<logical> Ptr>
     EVE_FORCEINLINE explicit logical(Ptr ptr) noexcept
-                  : storage_base(detail::load(eve::as_<logical>{}, ptr))
+                  : storage_base(load(ignore_none, safe, eve::as_<logical>{}, ptr))
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/arch/cpu/logical_wide.hpp:130]: Gonna remove those

File: include/eve/detail/function/simd/common/load.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/common/load.hpp:35]: Fine

File: include/eve/module/real/core/function/regular/generic/load.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/module/real/core/function/regular/generic/load.hpp:85]: Added laod(ptr,as) + variants and changed this

File: include/eve/detail/function/simd/arm/neon/load.hpp [hunk 11]
```diff
@@ -16,11 +16,18 @@
...
+  EVE_FORCEINLINE wide<T, N> load_( EVE_SUPPORTS(neon128_)
+                                  , ignore_none_ const&, safe_type const&
+                                  , eve::as_<wide<T, N>> const& tgt, Ptr ptr
+                                  )
+  requires simd_compatible_ptr<Ptr,wide<T, N>> && arm_abi<abi_t<T, N>>
   {
-    if constexpr( N::value * sizeof(T) >= arm_64_::bytes )
+    if constexpr( !std::is_pointer_v<Ptr> )
+    {
+      return load(ignore_none,safe,tgt, ptr.get());
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/neon/load.hpp:28]: Only on MSVC. I tried to make the compiler generate the proper intrinsic using builtin_assume to no avail.

File: include/eve/memory/pointer.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/memory/pointer.hpp:74]: Next PR is `structured_ptr<product_type>`

File: include/eve/detail/function/simd/arm/neon/load.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/neon/load.hpp:28]: Done for as_raw_pointer

File: include/eve/detail/function/simd/arm/neon/load.hpp [hunk 12]
```diff
@@ -57,15 +64,18 @@ namespace eve::detail
...
+  requires simd_compatible_ptr<aligned_ptr<U, Lanes>,wide<T, N>> && arm_abi<abi_t<T, N>>
   {
     auto ptr = p.get();
 
-    if constexpr( aligned_ptr<T const, Lanes>::alignment() < 16 )
+    if constexpr( aligned_ptr<U, Lanes>::alignment() < 16 )
     {
-      return load(ptr, tgt);
+      return load_(EVE_RETARGET(cpu_), ignore_none, safe, tgt, ptr);
     }
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/neon/load.hpp:79]: Done

File: include/eve/algo/ptr_iterator.hpp [hunk 13]
```diff
@@ -34,20 +35,21 @@ namespace eve::algo
...
-    }
 
-    friend wide_value_type tagged_dispatch( eve::tag::load_, unaligned_ptr_iterator self )
+    template< relative_conditional_expr C, decorator S, typename Pack>
+    friend auto tagged_dispatch ( eve::tag::load_, C const& c, S const& s
+                                , eve::as_<Pack> const& , unaligned_ptr_iterator self
+                                )
     {
-      return eve::load(self.ptr, cardinal{});
+      using tgt_t = typename Pack::template rescale<Cardinal>;
```
Comment (DenisYaroshevskiy) [inline include/eve/algo/ptr_iterator.hpp:51]: Hmm. I'm not sold on this.
Ideally it should be:

If there is no `cardinal`, pass a cardinal.

Hmm. Hmm. Leave as is, I'd have to think about this.

File: include/eve/detail/function/simd/common/load.hpp [hunk 14]
```diff
@@ -50,89 +171,91 @@ namespace eve::detail
       }
     );
 
-    return that.storage();
+    return that;
   }
 
   //================================================================================================
-  // Emulation
+  // Load from a range [b, e[
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/common/load.hpp:178]: Auch :) typo

File: include/eve/memory/pointer.hpp [hunk 15]
```diff
@@ -8,11 +8,71 @@
...
+#include <eve/detail/kumi.hpp>
 
 #include <concepts>
 
+namespace eve
+{
+  template<typename T, typename Enable=void>
+  struct pointer_traits
+  {
+    using value_type = void;
```
Comment (DenisYaroshevskiy) [inline include/eve/memory/pointer.hpp:20]: Why void?
I mean, let's just not do anything.

And then just require `pointer_traits::value_type`

**Review by DenisYaroshevskiy — COMMENTED** [review]
Looks good so far.

File: include/eve/algo/ptr_iterator.hpp (hunk 13 above)
Comment (jfalcou) [reply include/eve/algo/ptr_iterator.hpp:51]: Fixed

File: include/eve/detail/function/simd/common/load.hpp (hunk 14 above)
Comment (DenisYaroshevskiy) [reply include/eve/detail/function/simd/common/load.hpp:178]: You didn't actually fix the typo(

File: include/eve/module/real/core/function/regular/generic/load.hpp [hunk 16]
```diff
@@ -240,69 +139,33 @@ namespace eve::detail
...
-                  -> decltype(eve::load(ptr, N))
+  template<relative_conditional_expr C, data_source Ptr, typename Pack>
+  DISABLE_SANITIZERS Pack load_ ( EVE_SUPPORTS(cpu_)
+                                , C const& cond, unsafe_type const&
+                                , eve::as_<Pack> const& tgt, Ptr ptr
+                                ) noexcept
   {
-    return eve::unsafe(eve::load[eve::ignore_none])(ptr, N);
-  }
+    if constexpr(sanitizers_are_on && !std::same_as<C,ignore_none_> )
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/load.hpp:151]: Only sanitizers are on.

Ignore none will be ignore none in strlen case as well. But it needs to be unsafe.

Comment (jfalcou) [reply include/eve/module/real/core/function/regular/generic/load.hpp:151]: this was changed due to this remark of yours:

```
and not `ignore_none`
```
_Originally posted by @DenisYaroshevskiy in https://github.com/jfalcou/eve/pull/783#discussion_r652838314_

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**